### PR TITLE
Implemented GET functions for API

### PR DESCRIPTION
**Pull request by richardzeng1**
Implemented what=
- questions
- course_pairings
- user_info
- tas
- surveys
- survey_results

refactored php script to allow multiple queries and data processing before sending data back to the client.

### CODE REVIEW

File: error_check.php [hunk 1]
```diff
@@ -1,4 +1,62 @@
 <?php
+/**
+ * Checks if required fields are selected
+ *
+ * @param user_id The user requesting the course data
+ * @return String A sql command to get the requested data
+ */
+function get_courses($user_id){
+    //echo "SELECT courses.course_code, user_association.user_id, user_association.section_code FROM courses INNER JOIN user_association ON courses.course_code=user_association.course_id where user_id='$user_id'";
```
Comment (siefkenj) [inline error_check.php]: Comments like this should be deleted in a PR

File: error_check.php [hunk 2]
```diff
@@ -1,4 +1,62 @@
...
+    return "SELECT courses.course_code, user_association.user_id, user_association.section_code FROM courses INNER JOIN user_association ON courses.course_code=user_association.course_id where user_id='$user_id'";
```
Comment (siefkenj) [inline error_check.php]: You need to linewrap this. It's too hard to review as a single line.

File: error_check.php [hunk 3]
```diff
@@ -1,4 +1,62 @@
...
+
+/**
+ * Checks if required fields are selected
+ *
+ * @param user_id The user requesting the course data
+ * @param survey_instance_id The survey request to view results
+ * @return String A sql command to get the requested data
+ */
+function get_results(){
+    $user_id = $_GET["users"];
```
Comment (siefkenj) [inline error_check.php]: Utility functions should avoid accessing global variables at all costs. If you need `$user_id` and `$ta_id`, pass that in when you call this function.

File: statements.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
+<?php
```
Comment (siefkenj) [inline statements.php]: Each file should start with why it exists. That is, explain what types of functions live in that file and why you would call them.

File: statements.php [hunk 5]
```diff
@@ -0,0 +1,115 @@
+<!--
+This file contains functions that returns sql queries for GET requests.
+This file returned data based on data specified by API documentation.
+-->
```
Comment (siefkenj) [inline statements.php]: Why is this called `statements.php`? Use future tense: `returns` instead of `returns`. Does the *file* return data or does the file contain helper functions which return such data?

File: statements.php [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+/**
+ * Returns sql query to get user data. If include_photo is set to FALSE
+ * then photos are not returned. In any other case, photos are returned.
+ *
+ * @return String A sql command that returns unordered list of user data
+ */
+function get_user_info(){
+    $user_id = $_GET["user_id"];
```
Comment (siefkenj) [inline statements.php]: Don't access global variables inside function calls.

File: get_info.php [hunk 7]
```diff
@@ -1 +1,171 @@
...
+    }
+} catch (Exception $e) {
+    $result = HTTPStatus(400);
+    date_default_timezone_set('America/Toronto');
+    error_log(date("Y-m-d h:i:sa")." : ".$e->getMessage()."\n", 3, "errors.log");
+    print json_encode($result, JSON_PRETTY_PRINT);
+    exit();
+}
+
+// Handle SQL error
```
Comment (siefkenj) [inline get_info.php]: Are you handling a sql error, or are you attempting to get data via sql request?

File: get_info.php [hunk 8]
```diff
@@ -1 +1,171 @@
...
+    // Whitelist check
+    check_columns(array_keys($data['data']), $result['table'], 'delete');
+
+    $table = $result['table'];
+    $condition = $result['condition'];
+
+    return "DELETE FROM ta_feedback.$table WHERE $condition;";
+}
+
+function HTTPStatus($num) {
```
Comment (siefkenj) [inline get_info.php]: I don't like this name. Can you think of something more descriptive? This function returns an error object corresponding to an http status... `http_response_code_to_error` is a bit wordy, but it tells you what the function is doing.

File: get_info.php [hunk 9]
```diff
@@ -1 +1,159 @@
...
+        date("Y-m-d h:i:sa") . " : " . $e->getMessage() . "\n",
+        3,
+        "errors.log"
+    );
+    print json_encode($result, JSON_PRETTY_PRINT);
+    exit();
+}
+
+// Attempt to execute sql command and print response in json format.
+// If a sql error occurs, HTTP status 500 is returned.
```
Comment (siefkenj) [inline get_info.php]: You mean a JSON error object is returned.

File: get_query.php [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+/**
+ * Returns sql query to get user data. If include_photo is set to FALSE
+ * then photos are not returned. In any other case, photos are returned.
+ *
+ * @return String A sql command that returns unordered list of user data
+ */
+function get_user_info($user_id, $include_photo)
+{
+    $photo = ", users.photo";
+    if (!$include_photo) {
```
Comment (siefkenj) [inline get_query.php]: Is there a reason you don't reverse this? 

    $photo = "";
    if ($include_photo) {...

File: get_info.php [hunk 11]
```diff
@@ -1 +1,159 @@
-<?php?>
+<?php
+require 'db/config.php';
+require 'get_query.php';
```
Comment (siefkenj) [inline get_info.php]: I think `query_string_generators.php` would be a better name for this file.

File: get_query.php [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+ * This file contains functions that returns sql queries for GET requests
+ * according to the API documentation.
+ */
+
+/**
+ * Returns sql query that returns an unordered list of questions.
+ *
+ * @return String A sql command that returns an unorderedlist of questions.
+ */
+function get_questions()
```
Comment (siefkenj) [inline get_query.php]: Let's rename all these functions as `gen_query_**` so it is clear what they do.

File: get_query.php [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+ * are optional parameters.
+ *
+ * @param ta_list True if function call if for list of TAs
+ * @return String A sql command that returns an unorderedlist of questions.
+ */
+function get_course_associations($user_id, $column, $course_code, $term)
+{
+    $specify_course = "";
+    if ($course_code) {
+        $specify_course = "WHERE course_code='$course_code'";
```
Comment (siefkenj) [inline get_query.php]: We should be creating PDO suitable strings, not raw query strings. Raw query strings allow for injection attacks.  See https://stackoverflow.com/questions/60174/how-can-i-prevent-sql-injection-in-php

File: get_info.php [hunk 14]
```diff
@@ -1 +1,159 @@
...
+            break;
+
+        default:
+            throw new Exception("InvalidPage");
+            break;
+    }
+}
+
+/**
+ * This functions sets the parameters and calls get_course_associations and
```
Comment (siefkenj) [inline get_info.php]: What are **the** parameters?

File: get_info.php [hunk 15]
```diff
@@ -1 +1,159 @@
...
+ * returns the results.
+ *
+ * @param num The HTTP status code
+ * @return array containing the HTTP status of request
+ */
+function set_parameters($ta_list)
+{
+    $user_id = $_GET["user_id"];
```
Comment (siefkenj) [inline get_info.php]: Stop using globals inside functions.

Comment (richardzeng1) [reply get_info.php]: I used a global variable inside that function to improve code reuse. There are 2 cases that call that function and if I want to remove the global variable, it would need to go inside the handle_get function.

Comment (siefkenj) [reply get_info.php]: Keep global variables to the top scope and pass in any data you need as parameters. The top scope should be the only part of the code that cares about state. All the state should be normalized first, and then the data can be dispatched to the appropriate functions.

In general (though not in this project), we would be writing test cases for these individual functions. If you rely on global state, then it's very hard to test the functions. Further, they could not be used as module--reliance on globals actually prevents reuse.

File: get_info.php (hunk 11 above)
Comment (richardzeng1) [reply get_info.php]: Would all query generators be within this file? Would it be a good idea to split GET and POST queries into separate files? What about get_query_generators.php?

File: get_query.php [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+ *
+ * @return String A sql command that returns unordered list of user data
+ */
+function get_user_info($user_id, $include_photo)
+{
+    $photo = ", users.photo";
+    if (!$include_photo) {
+        $photo = "";
+    }
+    return "SELECT users.user_id, users.name$photo FROM users WHERE users.user_id='$user_id';";
```
Comment (DarrenZhang01) [inline get_query.php]: I do not think "users.user_id='$user_id';" would work. In the documentation, the parameter "user_id" in the url can be a list of users. In this case, we cannot directly insert the value of this key into the SQL statement without processing on it. Since I am doing unit testing on this API, it does not work as expected when the user_id is a list of users.

Comment (richardzeng1) [reply get_query.php]: No, that's not true. user_id is the person using the application. It is only 1 person and not a list.

Comment (DarrenZhang01) [reply get_query.php]: github
@richardzeng1

File: get_info.php [hunk 17]
```diff
@@ -1 +1,200 @@
...
+        case "user_info":
+            $include_photo = true;
+            if (isset($parameters["include_photo"])) {
+                if ($parameters["include_photo"] == 'false') {
+                    $include_photo = false;
+                }
+            }
+            $list_of_users = is_list_of_users($parameters["user_id"]);
+            return gen_query_user_info($include_photo, $list_of_users);
+            break;
```
Comment (siefkenj) [inline get_info.php]: You don't need break statements after returns.

File: get_info.php [hunk 18]
```diff
@@ -1 +1,200 @@
...
+        ($parameters["what"] == "course_pairings") ||
+        ($parameters["what"] == "tas")
+    ) {
+        if ($parameters["what"] == "course_pairings") {
+            if (isset($parameters["column"])) {
+                $stmt->bindParam(':column', $parameters["column"]);
+            }
+        } else {
+            $ta = 'ta';
+            $stmt->bindParam(':column', $ta);
```
Comment (siefkenj) [inline get_info.php]: Why did you set a variable to use it only once?

File: get_info.php [hunk 19]
```diff
@@ -1 +1,200 @@
...
+        date("Y-m-d h:i:sa") . " : " . $e->getMessage() . "\n",
+        3,
+        "errors.log"
+    );
+    print json_encode($result, JSON_PRETTY_PRINT);
+    exit();
+}
+
+// Attempt to execute sql command and print response in json format.
+// If a sql error occurs, JSON error object.
```
Comment (siefkenj) [inline get_info.php]: *return* a JSON error object

File: get_info.php [hunk 20]
```diff
@@ -1 +1,200 @@
...
+    $conn->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
+    $stmt = $conn->prepare($query);
+
+    bind_variables($stmt, $_GET);
+    $stmt->execute();
+
+    if ($method == "GET") {
+        //fetch all results
+        $fetched = $stmt->fetchAll(PDO::FETCH_ASSOC);
+        $result['TYPE'] = $_GET["what"] . "_package";
```
Comment (siefkenj) [inline get_info.php]: I'm not thrilled with this... `$_GET["what"]` could be anything...no attempt has been made to sanitize the parameters.

Comment (richardzeng1) [reply get_info.php]: I was going to implement sanity checks on the next commit. This PR included more than I anticipated such as SQL injection protection and sanity checks.

Comment (siefkenj) [reply get_info.php]: Okay. Fix the other things and you can commit. We'll do the sanitizing in another PR.

File: get_info.php (hunk 18 above)
Comment (richardzeng1) [reply get_info.php]: bindParam needs a reference. Passing in a string will cause an error.

Comment (siefkenj) [reply get_info.php]: Okay. Put a comment to that effect so it's not confusing.

Comment (richardzeng1) [reply get_info.php]: The new schema requires me to change this so it is no longer a problem.

**Comment by siefkenj** [comment]
Is this ready to merge yet?

**Comment by richardzeng1** [comment]
fixes #51

File: get_info.php [hunk 21]
```diff
@@ -1 +1,295 @@
...
+                'DATA' => $questions
+            );
+            return $question_package;
+        case "course_pairings":
+            if($parameters["column"]=="sections"){
+                $course_pairings = get_query_result(
+                    gen_query_course_pairings_section(),
+                    $bind_variables
+                );
+            }else{
```
Comment (siefkenj) [inline get_info.php]: Let's format our sourcecode....

Comment (richardzeng1) [reply get_info.php]: Sorry. Still a work in progress. I'll squash these 2 commits afterwords. Please don't merge.

**Comment by richardzeng1** [comment]
fixes #49

File: get_info.php [hunk 22]
```diff
@@ -1 +1,308 @@
...
+<?php
+require 'get_query_generators.php';
+header("Content-type: application/json");
+try {
+    $method = "";
+    if (isset($_SERVER['REQUEST_METHOD'])) {
+        $query = "";
+        $method = $_SERVER['REQUEST_METHOD'];
+        // Request stored as associative array
+        $data = json_decode(file_get_contents("php://input"), true);
```
Comment (F-Ting) [inline get_info.php]: Remove

File: get_info.php [hunk 23]
```diff
@@ -1 +1,308 @@
-<?php?>
+<?php
+require 'get_query_generators.php';
+header("Content-type: application/json");
+try {
+    $method = "";
+    if (isset($_SERVER['REQUEST_METHOD'])) {
+        $query = "";
```
Comment (F-Ting) [inline get_info.php]: remove

File: get_info.php [hunk 24]
```diff
@@ -1 +1,308 @@
...
+        throw new Exception($error);
+    }
+} catch (Exception $e) {
+    $result = set_http_response(400);
+    date_default_timezone_set('America/Toronto');
+    error_log(
+        date("Y-m-d h:i:sa") . " : " . $e->getMessage() . "\n",
+        3,
+        "errors.log"
+    );
```
Comment (F-Ting) [inline get_info.php]: error logging should be removed once we're ready for production,  add as comment.

File: get_info.php [hunk 25]
```diff
@@ -1 +1,308 @@
...
+    print json_encode($result, JSON_PRETTY_PRINT);
+    exit();
+}
+
+function get_query_result($query_string, $bind_variables)
```
Comment (F-Ting) [inline get_info.php]: document the function

File: get_info.php [hunk 26]
```diff
@@ -1 +1,308 @@
...
+{
+    require '../db/config.php';
```
Comment (F-Ting) [inline get_info.php]: figure out a way to import configs safely, add as comment.

File: get_info.php [hunk 27]
```diff
@@ -1 +1,308 @@
...
+    $method = "";
+    if (isset($_SERVER['REQUEST_METHOD'])) {
+        $query = "";
+        $method = $_SERVER['REQUEST_METHOD'];
+        // Request stored as associative array
+        $data = json_decode(file_get_contents("php://input"), true);
+        switch ($method) {
+            case "GET":
+                print json_encode(handle_get($_GET));
+                break;
```
Comment (F-Ting) [inline get_info.php]: add exit() here when data is returned

File: get_info.php [hunk 28]
```diff
@@ -1 +1,308 @@
...
+        $fetched = $stmt->fetchAll(PDO::FETCH_ASSOC);
+        return $fetched;
+    } catch (PDOException $e) {
+        $result = set_http_response(500);
+        date_default_timezone_set('America/Toronto');
+        error_log(
+            date("Y-m-d h:i:sa") . " : " . $e->getMessage() . "\n",
+            3,
+            "errors.log"
+        );
```
Comment (F-Ting) [inline get_info.php]: same as the above catch block

File: get_info.php [hunk 29]
```diff
@@ -1 +1,308 @@
...
+        print json_encode($result, JSON_PRETTY_PRINT);
+        exit();
+    }
+}
+
+/**
+ * This function handles GET requests and returns a sql query that gets the
```
Comment (F-Ting) [inline get_info.php]: this no longer returns a sql query, rather the data being sent back to client

File: get_info.php [hunk 30]
```diff
@@ -1 +1,308 @@
...
+ * required data for the what according to the API documentation.
+ *
+ * @throws InvalidPage The requested what does not exist
+ * @return sql query to the corresponding what parameter
```
Comment (F-Ting) [inline get_info.php]: same as above

File: get_info.php [hunk 31]
```diff
@@ -1 +1,308 @@
...
+            $survey_package = get_list_of_surveys(
+                $role[0],
+                $provided_survey,
+                $bind_variables,
+                true
+            );
+            return $survey_package;
+        default:
+            throw new Exception("InvalidPage");
+            break;
```
Comment (richardzeng1) [inline get_info.php]: Unreachable code

File: get_info.php [hunk 32]
```diff
@@ -1 +1,308 @@
...
+    }
+}
+
+/**
+ * Returns true if user_id is a list of users.
```
Comment (richardzeng1) [inline get_info.php]: Incorrect documentation

File: get_info.php [hunk 33]
```diff
@@ -1 +1,308 @@
...
+ * @param user_id Value of user_id parameters
+ * @return TRUE if user_id is a list
+ */
+function get_list_of_surveys(
```
Comment (richardzeng1) [inline get_info.php]: comment some of the implementation details

File: get_info.php [hunk 34]
```diff
@@ -1 +1,308 @@
...
+
+    $list_of_users = false;
+    if (sizeof($user_id_list) > 1) {
+        $list_of_users = true;
+    }
+    return $list_of_users;
+}
+
+/**
+ * This functions sets parameters for gen_query_course_associations based on
```
Comment (richardzeng1) [inline get_info.php]: function also sets parameters for gen_query_course_pairings_section

File: get_info.php [hunk 35]
```diff
@@ -1 +1,308 @@
...
+ * what parameter. Function returns the result of gen_query_course_associations.
+ *
+ * @param num The HTTP status code
+ * @return array containing the HTTP status of request
```
Comment (richardzeng1) [inline get_info.php]: parameters and return is incorrect. Should be
parameters an associative array containing parameter and values
returns CourseAssociations object according to the UML diagram

File: get_query_generators.php [hunk 36]
```diff
@@ -0,0 +1,336 @@
...
+ *
+ * @return String A sql command that returns an unorderedlist of questions.
+ */
+function gen_query_questions()
+{
+    return "SELECT question_id, answer_type, content FROM questions;";
+}
+
+/**
+ * Returns sql query that returns an unordered list of sections related to a ta.
```
Comment (richardzeng1) [inline get_query_generators.php]: this function can be extended to show all sections a instructor is registered for as well.

File: get_query_generators.php [hunk 37]
```diff
@@ -0,0 +1,336 @@
...
+ * For what=tas, term parameter sepcifies the term returned and course_code
+ * sepcifies the course. term and course_code parameters are optional.
+ *
+ * For what=course_pairings, column parameter is required. term and course_code
+ * are optional parameters.
+ *
+ * @param ta_list True if function call if for list of TAs
+ * @return String A sql command that returns an unorderedlist of questions.
+ */
+function gen_query_course_associations($course_code, $term, $is_ta)
```
Comment (richardzeng1) [inline get_query_generators.php]: rename to course_pairings. Course associations makes no sense for this. Stay consistent with API naming convention.

File: get_query_generators.php [hunk 38]
```diff
@@ -0,0 +1,336 @@
...
+/**
+ * Returns sql query that returns an unordered list of tas or instructors.
+ * If $ta_list is true, a list of tas is returned. If $ta_list is false,
+ * the returned list depends on column parameter in the url.
+ *
+ * For what=tas, term parameter sepcifies the term returned and course_code
+ * sepcifies the course. term and course_code parameters are optional.
+ *
+ * For what=course_pairings, column parameter is required. term and course_code
+ * are optional parameters.
```
Comment (richardzeng1) [inline get_query_generators.php]: need to update documentation according to current implementation.

File: get_query_generators.php [hunk 39]
```diff
@@ -0,0 +1,336 @@
...
+        "t1.department_name FROM ($instructor_for_courses) t1 JOIN ($sections) t2 " .
+        "ON t1.section_id=t2.section_id;";
+
+    return $return_query;
+}
+
+/**
+ * Returns sql query to get user data. If include_photo is set to FALSE
+ * then photos are not returned. In any other case, photos are returned.
+ *
```
Comment (richardzeng1) [inline get_query_generators.php]: Add parameter documentation.

File: get_query_generators.php [hunk 40]
```diff
@@ -0,0 +1,336 @@
...
+    if ($list_of_users) {
+        return "SELECT users.user_id, users.name$photo FROM users WHERE FIND_IN_SET(users.user_id, :user_id);";
+    }
+    return "SELECT users.user_id, users.name$photo FROM users WHERE users.user_id=:user_id;";
+}
+
+/**
+ * Returns sql query to get data for survey results.
+ * If no survey is select, then it returns all the results available for the
+ * selected ta.
```
Comment (richardzeng1) [inline get_query_generators.php]: Add documentation for parameters

File: get_query_generators.php [hunk 41]
```diff
@@ -0,0 +1,336 @@
...
+ *
+ * @return String A sql command to get the requested data
```
Comment (richardzeng1) [inline get_query_generators.php]: Bad description.

**Comment by F-Ting** [comment]
fixes #20
